### DISCUSSION TITLE
CLI doesn't exit when finished.

### DISCUSSION THREAD

**Discussion by Milner99** [winder/Universal-G-Code-Sender#2029] [General]
Hi,

Can any one help me understand why the CLI doesn't exit when it's finished sending a carve. Using Rasp Pi 4 to a TinyG over USB.

`./start-cli.sh --controller TinyG --port /dev/ttyUSB0 --baud 115200 --print-stream --file /var/www/UniversalGcodeSender/currentCarve.gcode`

This is the send command, .gcode is full of carve instructions and they get fed back to the terminal. Seeing a lot of processCommand errors like this one also:
```
Oct 31, 2022 1:55:01 PM com.willwinder.universalgcodesender.gcode.util.GcodeParserUtils processCommand
WARNING: An unknown gcode command was detected in: G28.2Z0
```
Any thoughts or help would be hugely appreciated. Thank you.

**Comment by breiler** [comment]
Can we have a sample gcode file?

**Reply by Milner99** [reply]
This is the GCode that has the issues.
[currentCarve.gcode.txt](https://github.com/winder/Universal-G-Code-Sender/files/9903435/currentCarve.gcode.txt)

**Reply by Milner99** [reply]
Hi @winder,

Very much appreciate you looking into this - have you had any time to do so? Happy to share any further info if I can help. 

ty.

**Reply by breiler** [reply]
I have set up my TinyG board, but I don't have homing available on it. 

The commands for homing and setting absolute position is not recognized by UGS gcode parser which basically just mean that we don't know how to visualize the commands, but they will be sent to the controller. So you can ignore the warning:
```
Oct 31, 2022 1:55:01 PM com.willwinder.universalgcodesender.gcode.util.GcodeParserUtils processCommand
WARNING: An unknown gcode command was detected in: G28.2Z0
```

I found a couple of issues with the CLI-program where it doesn't wait for the controller to be initialized properly before starting to send the file. I think I have a couple of fixes for that. I am not sure if that is what caused the CLI-program to not exit properly, but with the adjustments I made it will now send the file and exit properly. I'll let you know when my changes are merged and ready for testing.

By the way, cool to see 4th axis gcode programs, too bad that we only support XYZ-visualization for on tinyg controllers. 😢 
I gave it a test run on a FluidNC controller but there seems to be some issues with the visualization, so thanks for the sample gcode file!

**Reply by Milner99** [reply]
Great work. Thank you. Any view on the timeline for the merge?

**Reply by breiler** [reply]
It is available in the latest nightly build:
https://github.com/winder/Universal-G-Code-Sender#downloads

**Reply by Milner99** [reply]
Sadly, the build from a download today 2022-11-08, run on the command line from a raspberry Pi 4, using the command above and GCode above completes the send OK (bit slow) but then seems to hang without exiting.  The app has been sat/stalled after the M02 line which is the last command line in the GCode now for around an hour. I believe there is a blank line at the end of the GCode file, could that be causing an issue?

Is the app waiting on the TinyG for output? or is this a bug? 

Happy to discuss further and if I can help with information or even a screen share to debug please let me know.

**Reply by Milner99** [reply]
Hi, any further update on this?

Thank you
